### PR TITLE
Path: PathArray based on multiple bases, and choice of direction

### PR DESCRIPTION
**Pull request by jimzim111**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by jimzim111** [comment]
These changes are intended to help reduce tool travel time.  Discussion of the first one was at https://forum.freecadweb.org/viewtopic.php?f=15&t=58768&p=505517#p505517.

The first change allows PathArrays to be based on multiple Paths instead of just one.  It makes a copy of all of the bases in each arrayed position before moving on to the next position.

The second change allows swapping the direction in which copies are arrayed (X first instead of Y first).

**Review by Russ4262 — APPROVED** [review]
I built and tested.  Works as advertised.  Code changes look good.

File: src/Mod/Path/PathScripts/PathArray.py [hunk 1]
```diff
@@ -52,6 +54,12 @@ def __init__(self, obj):
                         "Path", "The number of copies in Linear 1D and Polar pattern")
         obj.addProperty("App::PropertyVector", "Centre",
                         "Path", "The centre of rotation in Polar pattern")
+        obj.addProperty("App::PropertyBool", "SwapDirection",
```
Comment (sliptonic) [inline src/Mod/Path/PathScripts/PathArray.py:57]: Please wrap all user-visible property descriptions in the QT_TRANSLATE_NOOP function so they get translated into other languages.   See 'pattern method' above.

File: src/Mod/Path/PathScripts/PathArray.py [hunk 2]
```diff
@@ -52,6 +54,12 @@ def __init__(self, obj):
...
+                        "Path", "Make copies in X direction before Y in Linear 2D pattern")
+        obj.addProperty("App::PropertyInteger", "JitterPercent",
+                        "Path", "Percent of copies to randomly offset")
+        obj.addProperty("App::PropertyVectorDistance", "JitterMagnitude",
+                        "Path", "Maximum random offset of copies")
         obj.addProperty("App::PropertyLink", "ToolController",
                         "Path", QtCore.QT_TRANSLATE_NOOP("App::Property", "The tool controller that will be used to calculate the path"))
```
Comment (sliptonic) [inline src/Mod/Path/PathScripts/PathArray.py:64]: Does the array really need its own TC reference?  Under what circumstances would the user change the TC in the array and not in the underlying operations?  If the op really needs a local copy but the user shouldn't change it, please hide the property from the user.

File: src/Mod/Path/PathScripts/PathArray.py [hunk 3]
```diff
@@ -67,6 +75,8 @@ def __setstate__(self, state):
         return None
 
     def setEditorProperties(self, obj):
+        obj.setEditorMode('JitterPercent', 0)
+        obj.setEditorMode('JitterMagnitude', 0)
```
Comment (sliptonic) [inline src/Mod/Path/PathScripts/PathArray.py:79]: Interesting. What's the use-case for a jitter?

File: src/Mod/Path/PathScripts/PathArray.py [hunk 4]
```diff
@@ -155,54 +166,92 @@ def rotatePath(self, path, angle, centre):
...
+
 
-            obj.ToolController = obj.Base.ToolController
+    def execute(self, obj):
+        if isinstance(obj.Base, list):
+            base = obj.Base
+        else:
+            base = [obj.Base]
+
+        if len(base)>0:
```
Comment (sliptonic) [inline src/Mod/Path/PathScripts/PathArray.py:183]: The consequenc if len(base) == 0 isn't visible until the end.  Looking ahead, nothing happens.  So, instead you should check for that case first and  and return if it's true, raising an error of necessary.  Currently all the subsequent code is indented and dependent on that check which is confusing.

File: src/Mod/Path/PathScripts/PathArray.py [hunk 5]
```diff
@@ -155,54 +166,92 @@ def rotatePath(self, path, angle, centre):
...
+            for b in base:
+                if not b.isDerivedFrom("Path::Feature"):
+                    return
+                if not b.Path:
+                    return
+                if not b.ToolController:
+                    return
+                if b.ToolController != obj.ToolController:
+                    # this may be important if Job output is split by tool controller
+                    PathLog.warning('Arrays of paths having different tool controllers are handled according to the tool controller of the first path.')
```
Comment (sliptonic) [inline src/Mod/Path/PathScripts/PathArray.py:194]: Not required for acceptance of this PR but if you're up for it, you could also check if the base feature is a dressup and allow for arrays of those as well.  Currently there's no way to make an array of dressed up operations.

File: src/Mod/Path/PathScripts/PathArray.py (hunk 1 above)
Comment (jimzim111) [reply src/Mod/Path/PathScripts/PathArray.py:57]: Will do

File: src/Mod/Path/PathScripts/PathArray.py (hunk 2 above)
Comment (jimzim111) [reply src/Mod/Path/PathScripts/PathArray.py:64]: I think that makes sense to control the Array's TC by the underlying operations.  Will hide.

File: src/Mod/Path/PathScripts/PathArray.py (hunk 3 above)
Comment (jimzim111) [reply src/Mod/Path/PathScripts/PathArray.py:79]: It's an easy way to create a hand-crafted appearance.  When you have a lot of the same cut repeated over and over with small variations it looks more natural and less mechanical.

**Comment by sliptonic** [comment]
Merging.
In the future, unless other people are basing future work on your specific branch, please rebase your PR on upstream master rather than merging those commits back to your branch.  Rebasing keeps the history cleaner

**Pull request merged**